File: app_mantenimiento/lib/services/maintenance_service.dart
import '../config/api_config.dart';
import '../models/maintenance/maintenance_request.dart';
import '../models/maintenance/maintenance_response.dart';
import 'api_service.dart';

class MaintenanceService {
  Future<List<MaintenanceResponse>> getMaintenancesByVehicle(
      int vehicleId) async {
    final data =
        await ApiService.get(ApiConfig.maintenancesByVehicle(vehicleId));
    final list = ApiPayload.decodeList(data);
    return list.map((e) => MaintenanceResponse.fromJson(e)).toList();
  }

  Future<MaintenanceResponse> createMaintenance(
      MaintenanceRequest request) async {
    final raw = await ApiService.post(
      ApiConfig.maintenance,
      body: request.toJson(),
    );
    return MaintenanceResponse.fromJson(ApiPayload.decodeMap(raw));
  }

  Future<MaintenanceResponse> updateMaintenance(
      int id, MaintenanceRequest request) async {
    final raw = await ApiService.put(
      ApiConfig.maintenanceById(id),
      body: request.toJson(),
    );
    return MaintenanceResponse.fromJson(ApiPayload.decodeMap(raw));
  }

  Future<void> deleteMaintenance(int id) async {
    await ApiService.delete(ApiConfig.maintenanceById(id));
  }
}

File: frontend/src/app/core/services/maintenance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { MaintenanceResponse } from '../../models/maintenance/maintenance-response';
import { MaintenanceRequest } from '../../models/maintenance/maintenance-request';

@Injectable({ providedIn: 'root' })
export class MaintenanceService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  getMaintenancesByVehicle(vehicleId: number): Observable<MaintenanceResponse[]> {
    return this.http.get<MaintenanceResponse[]>(
      `${this.apiUrl}/maintenance/vehicle/${vehicleId}`
    );
  }

  createMaintenance(data: MaintenanceRequest): Observable<MaintenanceResponse> {
    return this.http.post<MaintenanceResponse>(`${this.apiUrl}/maintenance`, data);
  }

  updateMaintenance(id: number, data: MaintenanceRequest): Observable<MaintenanceResponse> {
    return this.http.put<MaintenanceResponse>(
      `${this.apiUrl}/maintenance/${id}`, data
    );
  }

  deleteMaintenance(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/maintenance/${id}`);
  }
}
